resources/net.go

```diff
@@ -0,0 +1,726 @@
+		}
+	}
+	if obj.Gateway != "" {
+		if g := net.ParseIP(obj.Gateway); g == nil {
+			return fmt.Errorf("could not parse gateway: %s", obj.Gateway)
+		}
+	}
+
+	// validate the interface name
+	_, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+
+	return obj.BaseRes.Validate()
+}
+
+// Init runs some startup code for this resource.
+func (obj *NetRes) Init() error {
+	var err error
+	// get the network link's dbus path
+	obj.IfacePath, err = dbusIfacePath(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error getting iface dbus path")
+	}
+
+	// establish a go-systemd dbus connection
+	obj.sdbus, err = systemd.New()
+	if err != nil {
+		return errwrap.Wrapf(err, "failed to connect to systemd")
+	}
+	// establish a coreos godbus dbus connection
+	obj.gdbus, err = util.SystemBusPrivateUsable()
+	if err != nil {
+		return errwrap.Wrap(err, "error establishing system bus")
+	}
+
+	return obj.BaseRes.Init()
+}
+
+// Watch listens for networkd events by eavesdropping on the DBus Properties
+// interface (org.freedesktop.DBus.Properties) and filtering for signals from
+// the network device. It also watches the networkd service via a dbus unit
+// subscription, and the unit file responsible for configuring the interface.
+func (obj *NetRes) Watch() error {
+	var err error
+	// make sure systemd is running
+	if !systemdUtil.IsRunningSystemd() {
+		return fmt.Errorf("systemd is not running")
+	}
+
+	// create a dbus subscription to watch the networkd service unit
+	set := obj.sdbus.NewSubscriptionSet()
+	svcChan, svcErr := set.Subscribe() // svc event channel
+	set.Add(NetworkdService)
+
+	// Add a DBus rule to watch the "PropertiesChanged" signal and get messages
+	// from the network link's dbus path.
+	args := fmt.Sprintf(DBusEavesdropSprint, DBusPropertiesInterface, obj.IfacePath)
+	call := obj.gdbus.BusObject().Call(DBusAddMatch, 0, args)
+	if err := call.Err; err != nil {
+		return errwrap.Wrapf(err, "error creating dbus call")
+	}
+
+	// TODO: verify that implementation doesn't deadlock if there are unread
+	// messages left in the channel
+	ifaceChan := make(chan *dbus.Signal, 10) // iface event channel
+	defer close(ifaceChan)
+
+	obj.gdbus.Signal(ifaceChan)
+	defer obj.gdbus.RemoveSignal(ifaceChan)
+
+	// build the path to the configuration file
+	obj.UnitFilePath = NetworkdUnitFileDir + IfacePrefix + obj.GetName() + NetworkdUnitFileExt
+
+	// watch the configuration file
+	recWatcher, err := recwatch.NewRecWatcher(obj.UnitFilePath, false)
+	if err != nil {
+		return err
+	}
+	defer recWatcher.Close()
+
+	// notify engine that we're running
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false
+	var exit *error
+
+	for {
+		select {
+		case event, ok := <-ifaceChan:
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
```

Comment:
In general it's always nice if we can cause the whole "house" to come down, by asking the sender of this to shutdown, we process all it's events until it closes, and then when we get that close we shutdown too.

When we're receiving events from two sources (file and dbus) we ask them *both* to close, and when each closes, we do this:

```
if !ok {
    if someFlag {
        return nil
    }
    someFlag = true
    continue
}
```

This works for two senders. When the 2nd of the two shuts down, we exit :)

(For N senders, you need a `map` and you have to check to see if everyone's done each time someone closes. Remember this is the number of senders which we have a way to actually ask to exit! And if you're not expecting one to close ever then you error of course!)